In [1]:
# For tips on running notebooks in Google Colab, see
# https://docs.pytorch.org/tutorials/beginner/colab
%matplotlib inline

[Learn the Basics](intro.html) \|\|
[Quickstart](quickstart_tutorial.html) \|\|
[Tensors](tensorqs_tutorial.html) \|\| [Datasets &
DataLoaders](data_tutorial.html) \|\|
[Transforms](transforms_tutorial.html) \|\| [Build
Model](buildmodel_tutorial.html) \|\|
[Autograd](autogradqs_tutorial.html) \|\| **Optimization** \|\| [Save &
Load Model](saveloadrun_tutorial.html)

Optimizing Model Parameters
===========================

Now that we have a model and data it\'s time to train, validate and test
our model by optimizing its parameters on our data. Training a model is
an iterative process; in each iteration the model makes a guess about
the output, calculates the error in its guess (*loss*), collects the
derivatives of the error with respect to its parameters (as we saw in
the [previous section](autograd_tutorial.html)), and **optimizes** these
parameters using gradient descent. For a more detailed walkthrough of
this process, check out this video on [backpropagation from
3Blue1Brown](https://www.youtube.com/watch?v=tIeHLnjs5U8).

Prerequisite Code
-----------------

We load the code from the previous sections on [Datasets &
DataLoaders](data_tutorial.html) and [Build
Model](buildmodel_tutorial.html).


In [2]:
import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision.transforms import ToTensor

training_data = datasets.FashionMNIST(
    root="data",
    train=True,
    download=True,
    transform=ToTensor()
)

test_data = datasets.FashionMNIST(
    root="data",
    train=False,
    download=True,
    transform=ToTensor()
)

train_dataloader = DataLoader(training_data, batch_size=64)
test_dataloader = DataLoader(test_data, batch_size=64)

class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28*28, 512),
            nn.ReLU(),
            nn.Linear(512, 512),
            nn.ReLU(),
            nn.Linear(512, 10),
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits

model = NeuralNetwork()

100%|██████████| 26.4M/26.4M [00:01<00:00, 17.9MB/s]
100%|██████████| 29.5k/29.5k [00:00<00:00, 309kB/s]
100%|██████████| 4.42M/4.42M [00:00<00:00, 5.68MB/s]
100%|██████████| 5.15k/5.15k [00:00<00:00, 14.4MB/s]


In [3]:
for name, param in model.named_parameters():
    print(name, param.shape)


linear_relu_stack.0.weight torch.Size([512, 784])
linear_relu_stack.0.bias torch.Size([512])
linear_relu_stack.2.weight torch.Size([512, 512])
linear_relu_stack.2.bias torch.Size([512])
linear_relu_stack.4.weight torch.Size([10, 512])
linear_relu_stack.4.bias torch.Size([10])


Hyperparameters
===============

Hyperparameters are adjustable parameters that let you control the model
optimization process. Different hyperparameter values can impact model
training and convergence rates ([read
more](https://pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html)
about hyperparameter tuning)

We define the following hyperparameters for training:

:   -   **Number of Epochs** - the number of times to iterate over the
        dataset
    -   **Batch Size** - the number of data samples propagated through
        the network before the parameters are updated
    -   **Learning Rate** - how much to update models parameters at each
        batch/epoch. Smaller values yield slow learning speed, while
        large values may result in unpredictable behavior during
        training.


In [3]:
learning_rate = 1e-3
batch_size = 64
epochs = 5

Optimization Loop
=================

Once we set our hyperparameters, we can then train and optimize our
model with an optimization loop. Each iteration of the optimization loop
is called an **epoch**.

Each epoch consists of two main parts:

:   -   **The Train Loop** - iterate over the training dataset and try
        to converge to optimal parameters.
    -   **The Validation/Test Loop** - iterate over the test dataset to
        check if model performance is improving.

Let\'s briefly familiarize ourselves with some of the concepts used in
the training loop. Jump ahead to see the
`full-impl-label`{.interpreted-text role="ref"} of the optimization
loop.

Loss Function
-------------

When presented with some training data, our untrained network is likely
not to give the correct answer. **Loss function** measures the degree of
dissimilarity of obtained result to the target value, and it is the loss
function that we want to minimize during training. To calculate the loss
we make a prediction using the inputs of our given data sample and
compare it against the true data label value.

Common loss functions include
[nn.MSELoss](https://pytorch.org/docs/stable/generated/torch.nn.MSELoss.html#torch.nn.MSELoss)
(Mean Square Error) for regression tasks, and
[nn.NLLLoss](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html#torch.nn.NLLLoss)
(Negative Log Likelihood) for classification.
[nn.CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss)
combines `nn.LogSoftmax` and `nn.NLLLoss`.

We pass our model\'s output logits to `nn.CrossEntropyLoss`, which will
normalize the logits and compute the prediction error.


In [4]:
# Initialize the loss function
loss_fn = nn.CrossEntropyLoss()

Optimizer
=========

Optimization is the process of adjusting model parameters to reduce
model error in each training step. **Optimization algorithms** define
how this process is performed (in this example we use Stochastic
Gradient Descent). All optimization logic is encapsulated in the
`optimizer` object. Here, we use the SGD optimizer; additionally, there
are many [different
optimizers](https://pytorch.org/docs/stable/optim.html) available in
PyTorch such as ADAM and RMSProp, that work better for different kinds
of models and data.

We initialize the optimizer by registering the model\'s parameters that
need to be trained, and passing in the learning rate hyperparameter.


In [6]:
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

* `optimizer` 创建一个优化器对象= 负责“如何更新模型参数”的小助手。
* `torch.optim`是一个算法工具箱，包含 `Adam` `SGD` 等等
* `torch.optim.SGD`选择“随机梯度下降”这个优化算法
* `model.parameters()`返回一个 可迭代对象，里面包含了 模型中所有需要训练的参数

Inside the training loop, optimization happens in three steps:

:   -   Call `optimizer.zero_grad()` to reset the gradients of model
        parameters. Gradients by default add up; to prevent
        double-counting, we explicitly zero them at each iteration.
    -   Backpropagate the prediction loss with a call to
        `loss.backward()`. PyTorch deposits the gradients of the loss
        w.r.t. each parameter.
    -   Once we have our gradients, we call `optimizer.step()` to adjust
        the parameters by the gradients collected in the backward pass.


Full Implementation {#full-impl-label}
===================

We define `train_loop` that loops over our optimization code, and
`test_loop` that evaluates the model\'s performance against our test
data.


In [7]:
def train_loop(dataloader, model, loss_fn, optimizer):
    size = len(dataloader.dataset)
    # Set the model to training mode - important for batch normalization and dropout layers
    # Unnecessary in this situation but added for best practices
    model.train()
    for batch, (X, y) in enumerate(dataloader):
        # Compute prediction and loss
        pred = model(X)
        loss = loss_fn(pred, y)

        # Backpropagation
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

        if batch % 100 == 0:
            loss, current = loss.item(), batch * batch_size + len(X)
            print(f"loss: {loss:>7f}  [{current:>5d}/{size:>5d}]")


def test_loop(dataloader, model, loss_fn):
    # Set the model to evaluation mode - important for batch normalization and dropout layers
    # Unnecessary in this situation but added for best practices
    model.eval()
    size = len(dataloader.dataset)
    num_batches = len(dataloader)
    test_loss, correct = 0, 0

    # Evaluating the model with torch.no_grad() ensures that no gradients are computed during test mode
    # also serves to reduce unnecessary gradient computations and memory usage for tensors with requires_grad=True
    with torch.no_grad():
        for X, y in dataloader:
            pred = model(X)
            test_loss += loss_fn(pred, y).item()
            correct += (pred.argmax(1) == y).type(torch.float).sum().item()

    test_loss /= num_batches
    correct /= size
    print(f"Test Error: \n Accuracy: {(100*correct):>0.1f}%, Avg loss: {test_loss:>8f} \n")

* `model.train()` 告诉Pytorch 我要开始训练了，在此处可有可不有属于好习惯。
* `for batch, (X, y) in enumerate(dataloader):`
  `enumerate`会返回一个batch索引
  `DataLoader`是批次发生器，每次迭代给一个batch (例如64张图），每个batch的格式是（X， y）
  这里`enumerate(dataloader)`每次返回的内容是(batch_index, dataloader_output), 等同于(batch_index, (X, y))
* `train_dataloader = DataLoader(training_data, batch_size=64)` 这里指定了一次返回的样本数。所以到了training阶段每次for循环会返回64个样本。
---
关于后向传播过程：
1. Parameter是weight和bias，在定义Linear层的时候就自动生成了。
  执行 `nn.Linear`的时候会自动创建两个张量
  self.weight = nn.Parameter(...)
  self.bias   = nn.Parameter(...)
2. 执行 `model = NeuralNetwork()`的时候：
  PyTorch 扫描 model 里的所有属性
  找到所有 nn.Module 的子模块（Linear 就是）
  找到所有 nn.Parameter
  自动把它们加入到 model 的 parameter 列表里
3. `optimizer.step()` 是优化器，执行更新。
    `loss.backward()`执行时会得到 `param.grad = 某个梯度值`
    优化器接着执行：
    param.data = param.data - lr * param.grad （也就是SGD）
---

当你在模型里写下 `nn.Linear(784, 512)` 这一层时，PyTorch 会在内部自动为这层创建对应的可训练参数，也就是形状为 `[512, 784]` 的 `weight` 和形状为 `[512]` 的 `bias`。随后，当你实例化整个网络（`model = NeuralNetwork()`）时，PyTorch 会把这些用 `nn.Parameter` 包装过的张量自动注册到模型里，使它们出现在 `model.parameters()` 中，方便优化器统一管理。在训练过程中，每当你执行 `pred = model(X)` 做前向传播时，这些 `weight` 和 `bias` 就会参与计算，生成模型对当前输入的预测；接着，你根据预测和真实标签计算出 `loss`，调用 `loss.backward()` 时，PyTorch 会沿着计算图自动进行反向传播，为每个参数计算梯度，并写入对应的 `weight.grad` 和 `bias.grad` 中。然后，当你调用 `optimizer.step()` 时，优化器会读取这些梯度，用类似 `weight -= lr * weight.grad`、`bias -= lr * bias.grad` 这样的规则去更新参数的数值，从而让模型朝着减小损失的方向“学习”。更新完成后，再通过 `optimizer.zero_grad()` 将所有参数的梯度清空，避免上一批数据的梯度残留影响下一批的计算，如此一批一批循环，模型的参数就会在不断前向、反向、更新、清梯度的过程中逐渐被优化。



* test部分
1. model.eval()：进入“评估模式”
  你的模型没有这些层，但依然推荐使用，保持良好习惯。
  📌 eval() 不会关闭梯度，只是把模型切换为“不随机、不变化”的状态。
2. `with torch.no_grad():`
  关闭梯度计算。
  **为什么要特意关闭**：
  不写 no_grad，PyTorch 会构建计算图（非常耗时）；
  即使你不调用 backward()，PyTorch 默认会：记录前向传播的计算步骤；为自动求导构建计算图；为每一步保留中间结果（activation）。

We initialize the loss function and optimizer, and pass it to
`train_loop` and `test_loop`. Feel free to increase the number of epochs
to track the model\'s improving performance.


In [8]:
loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

epochs = 10
for t in range(epochs):
    print(f"Epoch {t+1}\n-------------------------------")
    train_loop(train_dataloader, model, loss_fn, optimizer)
    test_loop(test_dataloader, model, loss_fn)
print("Done!")

Epoch 1
-------------------------------
loss: 2.314596  [   64/60000]
loss: 2.292724  [ 6464/60000]
loss: 2.274892  [12864/60000]
loss: 2.261275  [19264/60000]
loss: 2.248711  [25664/60000]
loss: 2.228521  [32064/60000]
loss: 2.237839  [38464/60000]
loss: 2.204458  [44864/60000]
loss: 2.200596  [51264/60000]
loss: 2.167033  [57664/60000]
Test Error: 
 Accuracy: 37.2%, Avg loss: 2.158156 

Epoch 2
-------------------------------
loss: 2.171212  [   64/60000]
loss: 2.161684  [ 6464/60000]
loss: 2.107681  [12864/60000]
loss: 2.120569  [19264/60000]
loss: 2.069658  [25664/60000]
loss: 2.017059  [32064/60000]
loss: 2.045573  [38464/60000]
loss: 1.964704  [44864/60000]
loss: 1.972304  [51264/60000]
loss: 1.899433  [57664/60000]
Test Error: 
 Accuracy: 57.3%, Avg loss: 1.895734 

Epoch 3
-------------------------------
loss: 1.928349  [   64/60000]
loss: 1.904098  [ 6464/60000]
loss: 1.788886  [12864/60000]
loss: 1.823181  [19264/60000]
loss: 1.711145  [25664/60000]
loss: 1.667832  [32064/600

Further Reading
===============

-   [Loss
    Functions](https://pytorch.org/docs/stable/nn.html#loss-functions)
-   [torch.optim](https://pytorch.org/docs/stable/optim.html)
-   [Warmstart Training a
    Model](https://pytorch.org/tutorials/recipes/recipes/warmstarting_model_using_parameters_from_a_different_model.html)
